test/Hessian_test.cpp

```diff
@@ -0,0 +1,433 @@
+/*******************************************************************************
+
+   Tests for hessian methods.
+
+   This file is part of XAD, a comprehensive C++ library for
+   automatic differentiation.
+
+   Copyright (C) 2010-2024 Xcelerit Computing Ltd.
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU Affero General Public License as published
+   by the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Affero General Public License for more details.
+
+   You should have received a copy of the GNU Affero General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+******************************************************************************/
+
+#include <XAD/Hessian.hpp>
+#include <XAD/XAD.hpp>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <complex>
+#include <functional>
+#include <list>
+#include <type_traits>
+#include <typeinfo>
+#include <vector>
+
+using namespace ::testing;
+
+TEST(HessianTest, QuadraticForwardAdjoint)
+{
+    typedef xad::fwd_adj<double> mode;
+    typedef mode::active_type AD;
+    typedef mode::tape_type tape_type;
+
+    tape_type tape;
+
+    std::vector<AD> x = {3.0, 2.0};
+
+    // f(x) = x[0]^2 + x[1]^2
+    auto foo = [](std::vector<AD> &x) -> AD { return x[0] * x[0] + x[1] * x[1]; };
+
+    std::vector<std::vector<AD>> expected_hessian = {{2.0, 0.0}, {0.0, 2.0}};
+
+    auto computed_hessian = xad::computeHessian<double>(x, foo, &tape);
+
+    for (unsigned int i = 0; i < expected_hessian.size(); i++)
+        for (unsigned int j = 0; j < expected_hessian[i].size(); j++)
+            ASSERT_EQ(expected_hessian[i][j], computed_hessian[i][j]);
```

Comment:
Floating point comparison for equality is dangerous due to tiny precision issues. And the gmock matcher based assertions produce easier / better error messages on test failures. So please use this:

```suggestion
    for (unsigned int i = 0; i < expected_hessian.size(); i++)
        for (unsigned int j = 0; j < expected_hessian[i].size(); j++)
            EXPECT_THAT(computed_hessian[i][j], DoubleEq(expected_hessian[i][j]));
```

if you feel adventurous, you can also use this ;-):

```c++
    for (unsigned int i = 0; i < expected_hessian.size(); i++)
        EXPECT_THAT(computed_hessian[i], PointWise(DoubleEq(), expected_hessian[i]);
```

And please use the `DoubleEq` for all floating point comparisons in the other tests.